XenAdmin/Dialogs/EnablePvsReadCachingDialog.cs

```diff
@@ -0,0 +1,184 @@
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND 
+ * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+ * SUCH DAMAGE.
+ */
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Drawing;
+using System.Linq;
+using XenAdmin.Controls;
+using XenAdmin.Core;
+using XenAPI;
+using XenAdmin.Actions;
+
+
+namespace XenAdmin.Dialogs
+{
+    public partial class EnablePvsReadCachingDialog : XenDialogBase
+    {
+        private IList<VM> vms;
+
+        public EnablePvsReadCachingDialog(IList<VM> vms)
+        {
+            InitializeComponent();
+
+            this.vms = vms;
+
+            if (vms.Count() > 1)
+            {
+                rubricLabel.Text = Messages.ENABLE_PVS_READ_CACHING_RUBRIC_MULTIPLE;
+            }
+            else
+            {
+                rubricLabel.Text = Messages.ENABLE_PVS_READ_CACHING_RUBRIC_SINGLE;
+            }
+
+            PopulateSiteList();
+        }
+
+        private void PopulateSiteList()
+        {
+            // We assume all VMs share a pool
+            var vm = vms[0];
+            foreach (var site in vm.Connection.Cache.PVS_sites)
+            {
+                var siteToAdd = new ToStringWrapper<PVS_site>(site, site.name_label);
+                pvsSiteList.Items.Add(siteToAdd);
+            }
+
+            if (pvsSiteList.Items.Count > 0)
+            {
+                pvsSiteList.SelectedIndex = 0;
+            }
+        }
+
+        private void enableButton_Click(object sender, EventArgs e)
+        {
+            var siteItemSelected = (ToStringWrapper<PVS_site>) pvsSiteList.SelectedItem;
+            var siteSelected = siteItemSelected.item;
+
+            var actions = new List<AsyncAction>();
+
+            foreach (var vm in vms)
+            {
+                var action = GetAsyncActionForVm(vm, siteSelected);
+                if (action != null)
+                {
+                    actions.Add(action);
+                }
+            }
+
+            if (actions.Any())
+            {
+                var session = vms[0].Connection.Session;
+
+                if (actions.Count == 1)
+                {
+                    actions[0].RunExternal(session);

```

Comment:
We shoul call RunAsync() instead of RunExternal(), because RunExternal() runs the action non-async, so the thread is blocked until the action is completed. Same applies for ParallelAction below.